packages/beacon-node/src/sync/utils/downloadByRange.ts

```diff
@@ -0,0 +1,886 @@
+      },
+      "Must request blobs for blob-only forks"
+    );
+  }
+
+  if (isForkPostFulu(forkName) && !columnsRequest) {
+    throw new DownloadByRangeError(
+      {
+        code: DownloadByRangeErrorCode.MISSING_COLUMNS_REQUEST,
+        slotRange,
+      },
+      "Must request columns for forks with columns"
+    );
+  }
+
+  if (blocksRequest.startSlot !== dataRequest.startSlot) {
+    throw new DownloadByRangeError({
+      code: DownloadByRangeErrorCode.START_SLOT_MISMATCH,
+      blockStartSlot: blocksRequest.startSlot,
+      dataStartSlot: dataRequest.startSlot,
+    });
+  }
+
+  if (blocksRequest.count !== dataRequest.count) {
+    throw new DownloadByRangeError({
+      code: DownloadByRangeErrorCode.COUNT_MISMATCH,
+      blockCount: blocksRequest.count,
+      dataCount: dataRequest.count,
+    });
+  }
+
+  return slotRange;
+}
+
+/**
+ * Should not be called directly. Only exported for unit testing purposes
+ */
+export async function requestByRange({
+  network,
+  peerIdStr,
+  blocksRequest,
+  blobsRequest,
+  columnsRequest,
+}: DownloadByRangeRequests & {
+  network: INetwork;
+  peerIdStr: PeerIdStr;
+}): Promise<DownloadByRangeResponses> {
+  let blocks: undefined | SignedBeaconBlock[];
+  let blobSidecars: undefined | deneb.BlobSidecars;
+  let columnSidecars: undefined | fulu.DataColumnSidecars;
+
+  const requests: Promise<unknown>[] = [];
+
+  if (blocksRequest) {
+    requests.push(
+      network.sendBeaconBlocksByRange(peerIdStr, blocksRequest).then((blockResponse) => {
+        blocks = blockResponse.map(({data}) => data);
+      })
+    );
+  }
+
+  if (blobsRequest) {
+    requests.push(
+      network.sendBlobSidecarsByRange(peerIdStr, blobsRequest).then((blobResponse) => {
+        blobSidecars = blobResponse;
+      })
+    );
+  }
+
+  if (columnsRequest) {
+    requests.push(
+      network.sendDataColumnSidecarsByRange(peerIdStr, columnsRequest).then((columnResponse) => {
+        columnSidecars = columnResponse;
+      })
+    );
+  }
+
+  await Promise.all(requests);
+
+  return {
+    blocks,
+    blobSidecars,
+    columnSidecars,
+  };
+}
+
+/**
+ * Should not be called directly. Only exported for unit testing purposes
+ */
+export function validateResponses({
+  peerIdStr,
+  slotRangeString,
+  blocksRequest,
+  blobsRequest,
+  columnsRequest,
+  blocks,
+  blobSidecars,
+  columnSidecars,
+}: DownloadByRangeRequests & DownloadByRangeResponses & {peerIdStr: string; slotRangeString: string}): void {
+  if (!blocks) {
+    throw new DownloadByRangeError(
+      {
+        code: DownloadByRangeErrorCode.MISSING_BLOCKS_RESPONSE,
+        slotRange: slotRangeString,
+      },
+      "No blocks to validate requests against"
+    );
+  }
+  if (!blocksRequest) {
+    throw new DownloadByRangeError(
+      {
+        code: DownloadByRangeErrorCode.MISSING_BLOCKS_RESPONSE,
+        slotRange: slotRangeString,
+      },
+      "No blocks request to validate requests against"
+    );
+  }
```

Comment:
Why is `blocksRequest` optional if it will result in an error unconditionally? Just make it required if it always will exist (and change Batch.request to avoid the undefined `blocksRequest` case)

Reply:
This was the bug I found that needs the blocks passed back in from the Batch.  I realized there could be a round of requests where blocks are already present so I updated the types to break the build so I wouldn't leave the bug behind.  This is the part I mentioned needed to be refactored